ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+      } else {
+        responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
+        chunkOperationTracker.onResponse(getRequestInfo.replicaId, false);
+      }
+    }
+
+    /**
+     * Process the given {@link ServerErrorCode} and set operation status accordingly.
+     * @param errorCode the {@link ServerErrorCode} to process.
+     */
+    void processServerError(ServerErrorCode errorCode) {
+      setChunkException(new RouterException("Server returned: " + errorCode, RouterErrorCode.UnexpectedInternalError));
+    }
+
+    /**
+     * Set the exception associated with this chunk operation.
+     * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} error overrides any other
+     * previously received exception.
+     * @param exception the {@link RouterException} to possibly set.
+     */
+    void setChunkException(RouterException exception) {
+      if (chunkException == null || exception.getErrorCode() == RouterErrorCode.BlobDeleted
+          || exception.getErrorCode() == RouterErrorCode.BlobExpired) {
+        chunkException = exception;
+      }
+    }
+
+    /**
+     * @return true if this GetChunk is free so a chunk of the overall blob can be assigned to it.
+     */
+    boolean isFree() {
+      return state == ChunkState.Free;
+    }
+
+    /**
+     * @return true if this GetChunk is assigned with a chunk of the overall blob and is ready to issue requests.
+     */
+    boolean isReady() {
+      return state == ChunkState.Ready;
+    }
+
+    /**
+     * @return true if the operation on this chunk is in progress.
+     */
+    boolean isInProgress() {
+      return state == ChunkState.InProgress;
+    }
+
+    /**
+     * @return true if the operation on the current chunk is complete.
+     */
+    boolean isComplete() {
+      return state == ChunkState.Complete;
+    }
+  }
+
+  /**
+   * Special GetChunk used to retrieve and hold the first chunk of a blob. The first chunk is special because it
+   * could either be a metadata chunk of a composite blob, or the single chunk of a simple blob,
+   * and whether a chunk is composite or simple can only be determined after the first chunk is fetched.
+   */
+  private class FirstGetChunk extends GetChunk {
+    /**
+     * Construct a FirstGetChunk and initialize it with the given {@link BlobId}.
+     * @param blobId the {@link BlobId} to assign to this chunk. This will be the id of the overall blob.
+     */
+    FirstGetChunk(BlobId blobId) {
+      super(-1, blobId);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br>
+     * Post completion cleanup for the FirstGetChunk only involves clearing the correlation id map (so that
+     * subsequent responses are ignored). The rest of the state of the FirstGetChunk is never cleared.
+     */
+    @Override
+    void postCompletionCleanup() {
+      correlationIdToGetRequestInfo.clear();
+    }
+
+    @Override
+    GetOptions getGetOptions() {
+      return GetOptions.None;
```

Comment:
Shouldn't this is passed in arg to GetBlobOperation. In future, admin will be used to retrieve expired blobs, deleted blobs and so on which will use the router to fetch blobs